### ISSUE TITLE
Improve workflow for adding projects

### ISSUE THREAD

**Issue by 2color** [ipfs/ecosystem-directory#241]
Currently we rely on Airtable as the source of truth for projects in this repo.

The current workflow involves:
1. Reviewing the submissions on Airtable
1. Once the approval checkbox is checked the `json` filed in Airtable is populated with the content from all the fields
1. The JSON is manually copied into a file in this repo [described here](https://github.com/ipfs/ecosystem-directory#transforming-project-data)

## Challenges with this approach
- Moving the JSON from Airtable is error prone and requires opening a PR in this repository that also adds the images
- Making changes requires updating Airtable, having the JSON regenerated and copying over the changes here which is also 
- It's not possible to preview how the website looks with new submissions in Airtable

## Improvement ideas 
Two different ideas on how we can improve the workflow to reduce friction 

1. Move the submission process directly to this repository. Have a PR template so that submissions can be submitted and previewed directly here.
1. Have a script to sync between Airtable (using [their REST API](https://airtable.com/api)) and the files in this repository to avoid manually copying JSON. Could probably be in a GitHub action that routinely creates PRs for projects that have been approved/checked.

**Comment by orvn** [comment]
Great points @2color. Thanks for taking the time to outline these! I've inserted a few of my thoughts below.

> Do we still have a case/need for using Airtable? The main reason to use Airtable seems to be to allow non-developers to edit content. This comes with a fair bit of additional complexity which could potentially be avoided.

I'm not personally attached to it, but it depends on who will be maintaining the directories: the Airtable's original purpose was to have a user-friendly area to manage the content before shipping it to production.

> Move the submission process directly to this repository. Have a PR/project.json template so that submissions can be submitted and previewed directly here. Could be combined with a tool like https://forestry.io/ for easier editing of content.

I'm on board, and I do like Forestry. One issue might be that it doesn't provide a tabular view of content and each entry requires _a lot_ of click depth to get to.

> Have a script to sync between Airtable (using [their REST API](https://airtable.com/api)) and the files in this repository to avoid manually copying JSON. Could probably be in a GitHub action that routinely creates PRs for projects that have been approved/checked.

We can also have Airtable create the PR with keys and logos, on pressing a button?

> When submissions are made to Airtable, there's no easy way to ask the submitters for changes. This puts the burden of editing on us rather than being able to provide feedback and make changes in a single place.

This is one issue that can't be solved unless the Airtable base's view is somewhat public afaik. Once the Airtable form submission is made, it's immutable to the originator (this is by design, from Airtable).

**Comment by bmann** [comment]
I've used Forestry, if it can be added for edits for other elements of the site, so non Github users can do it, that would be great -- but the forms etc need a lot of work to be super usable. The team is working on a v2 product, Tina https://tina.io/ -- I haven't tried it yet.

It seems like the Airtable automation to create a PR might be the useful middle path?

And: a template could still be a good idea -- could also be an issue template? Still a manual step to include/PR/merge.

If the path is then "after you've submitted, you can make PRs to the repo" -- I don't think it's that much of a burden. For orgs that can't do this in house, they could also file an issue requesting updates. (that's still a lot of Github centric stuff).

I'm glad that I know this, because I actually have to edits to Fission's entry :)

**Comment by orvn** [comment]
> The team is working on a v2 product, Tina https://tina.io/ -- I haven't tried it yet.

I've tried Tina @bmann and it's pretty great actually! It's more intended for true visual content management of marketing websites. But the real blocker is that it only works with a really specific stack (NextJS). 

[Not sure yet](https://github.com/tinacms/tinacms/issues/1259) if expanding to other similar stacks is on their roadmap.

**Comment by mishmosh** [comment]
Sorry for the slow response. I think we do need to keep the data source in Airtable for 2 major reasons. We recut the data in multiple ways for other reporting & analysis in the ecosystem WG.

Very supportive of improving the automation to sync data from Airtable to json, though.

**Comment by nikipdia** [comment]
Offering my help to ease anything Airtable related, if needed. There would be workarounds to get things done in Airtable (Zapier integrations, collecting data elsewhere and importing to Airtable etc.) I am not great at Airtable scripts, and write to website from Airtable though.

`When submissions are made to Airtable, there's no easy way to ask the submitters for changes. This puts the burden of editing on us rather than being able to provide feedback and make changes in a single place.`

You can collect info in a Google form, write an automation to create and update records in Airtable when there are new record entries and updates in the Google sheet.

**Comment by orvn** [comment]
There's an Airtable automation that's running already, it can be enhanced to push directly to this repo (a branch that would get PR'd into `main`).

> You can collect info in a Google form, write an automation to create and update records in Airtable when there are new record entries and updates in the Google sheet.

Oh, that's a good idea. The Airtable would just need to be kept in good lockstep with the Google Form (e.g., fields like this would have to be consistent)

<img width="400" alt="fields-airtable" src="https://user-images.githubusercontent.com/3689720/183492713-a28df5a2-7579-4f51-930c-da8c1af4c6a9.png">

**Comment by nikipdia** [comment]
`The Airtable would just need to be kept in good lockstep with the Google Form (e.g., fields like this would have to be consistent)`
An [add on](https://workspace.google.com/marketplace/app/dynamic_fields/708020008404) solve this problem too.

**Comment by jamespohalloran** [comment]
> > The team is working on a v2 product, Tina https://tina.io/ -- I haven't tried it yet.
> 
> I've tried Tina @bmann and it's pretty great actually! It's more intended for true visual content management of marketing websites. But the real blocker is that it only works with a really specific stack (NextJS).
> 
> [Not sure yet](https://github.com/tinacms/tinacms/issues/1259) if expanding to other similar stacks is on their roadmap.

Tina Product Manager here! We're actively working on support for non-Next-based sites now actually!

**Comment by orvn** [comment]
> Tina Product Manager here! We're actively working on support for non-Next-based sites now actually!
That's amazing @jamespohalloran, thanks for the reply! ❤️

**Comment by orvn** [comment]
> Tina Product Manager here! We're actively working on support for non-Next-based sites now actually!

That's amazing @jamespohalloran! Thanks for the reply ❤️

**Comment by 2color** [comment]
Thanks for all the input.

Based on the fact that we need to keep submissions in Airtable for the reasons Mosh stated, I would suggest the following course of action:
- Keep Airtable as the canonical source of truth
- Create a **pull** script to fetch the records from Airtable here. The reason for pull rather than having push automation on the Airtable side is manifold:
  -  Since the goal of this automation is to update the website, it should live with the website
  - We can leverage GitHub Actions to automate as we desire while leveraging all the preexisting tooling for PR creation and workflow management.
  - We can leverage the code/sync script to also manually batch pull submissions if necessary or even detect deviations.
  
 I would suggest avoiding introducing Google Forms/Sheets into this as it would introduce unnecessary complexity.

This still doesn't address how we provide feedback to submitters on the copy to be published besides using email. But, having gone through some of the published submissions, they don't contain _that much_ copy, so I think we can start with the syncing and address the feedback challenge later if it's a problem.

**Comment by orvn** [comment]
> Create a pull script to fetch the records from Airtable here

Yeah, that's a really good point and this is the better pattern 🤩. Github Actions should be able to accomplish this (by running a shell or node script in the CI, and adding the Airtable token as an env in Github).

I'm going to be implementing something very similar (Github → Airtable) on another PL project in the next couple of weeks and can probably port over some of that for this project.

**Comment by 2color** [comment]
Closing this as its resolved by https://github.com/ipfs/ecosystem-directory/pull/249
